libcxx/include/__algorithm/fold.h

```diff
@@ -0,0 +1,118 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___ALGORITHM_FOLD_H
+#define _LIBCPP___ALGORITHM_FOLD_H
+
+#include <__concepts/assignable.h>
+#include <__concepts/convertible_to.h>
+#include <__concepts/invocable.h>
+#include <__concepts/movable.h>
+#include <__config>
+#include <__functional/invoke.h>
+#include <__functional/reference_wrapper.h>
+#include <__iterator/concepts.h>
+#include <__iterator/iterator_traits.h>
+#include <__iterator/next.h>
+#include <__ranges/access.h>
+#include <__ranges/concepts.h>
+#include <__ranges/dangling.h>
+#include <__type_traits/decay.h>
+#include <__type_traits/invoke.h>
+#include <__utility/forward.h>
+#include <__utility/move.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 23
+
+namespace ranges {
+template <class _Ip, class _Tp>
+struct in_value_result {
+  _Ip in;
+  _Tp result;
+};
+
+template <class _Ip, class _Tp>
+using fold_left_with_iter_result = in_value_result<_Ip, _Tp>;
+
+template <class _Fp, class _Tp, class _Ip, class _Rp, class _Up = decay_t<_Rp>>
+concept __indirectly_binary_left_foldable_impl =
+    convertible_to<_Rp, _Up> &&                    //
```

Comment:
This `convertible_to` is at the [bottom of _`indirectly-binary-left-foldable`_](https://eel.is/c++draft/algorithm.syn#concept:indirectly-binary-left-foldable:~:text=convertible_to%3Cinvoke_result_t%3CF%26%2C%20T%2C%20iter_reference_t%3CI%3E%3E%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20decay_t%3Cinvoke_result_t%3CF%26%2C%20T%2C%20iter_reference_t%3CI%3E%3E%3E%3E). I moved it into `__indirectly_binary_left_foldable_impl` to simplify the concept so we don't have to spell out `invoke_result_t<...>` more than once, and [asked it to be updated in the synopsis too](https://github.com/cplusplus/draft/pull/6734). The position of this concept's evaluation hasn't changed.

Reply:
For the room: Chris and I spent 20 minute discussing this offline.

OK, I don't see any value in the refactoring here. Practically speaking it gets us nothing.

However,, writing it out verbatim makes the code trivial to inspect for correctness. You simply diff it against the standard. 

As written this code is significantly harder to understand in the light of the subtleties of concept subsumption.

If I find time I'll be writing some docs for review that suggest we _always_ copy concepts given as exposition verbatim from the standard. Isn't that the whole reason they're there. 

I won't block this review on it. So I'm going to leave this conversation here so we can have it in a more appropriate venue.

Reply:
FWIW (and it's probably not much), as I was reviewing this PR, the difference between Chris' awesome implementation and the way that it was written in the standard did force me to spend a significant amount of time determining that Chris was following the standard correctly but his implementation was just smarter than I could comprehend.

In other words, I agree that it would be valuable (for reviewers, if nothing else) for these concepts to be carried verbatim from the spec, if possible.

Reply:
FWIW, I'm very much in favor of copying verbatim from the Standard unless there's a strong reason not to.

Reply:
I agree that we need to be copying verbatim for user-facing concepts. Exposition-only concepts are a different story because:

* they're only present for the purposes of conveying the necessary constraints without needing to open a _Constraints_ element;
* are optimised for communicating information in standardese, not for real-world software engineering;
* can be refactored editorially, meaning that they can silently change and we'd be out of sync without realising

Had this been a _Constraints_ element instead of a concept, we would've needed to construct the concept from prose, and it would look as it does here. I think it would be in our users' best interests to treat exposition-only concepts as a shorthand for a _Constraints_ element instead of something that's set in stone. In this case, it only mildly improves readability, and so it's not really worth debating, but there are other cases where it can actually matter (such as _`indirectly-binary-right-foldable`_, but we can have that discussion when I implement that).

Reply:
Can you please specify in some way that this is _not_ copied verbatim? That way nobody has to find this conversation to figure out if it was intentional?

Reply:
Done.